### PR TITLE
High Saliency Pointing Game

### PR DESCRIPTION
**Pull request by natalie-isak**
This PR adds in a new evaluation metric for explainability techniques. 

As part of this metric, the saliency scores for an object's image will be calculated, and then any scores below a given threshold (set by the user) are set to -1. The user can visualize this saliency map and can also calculate the percent of highly salient pixels that overlap with the gt bounding box.

### CODE REVIEW

File: python/vision_explanation_methods/explanation_evaluation/pointing_game.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        :param gt_bbox: bounding box for ground truth prediction
+        :type gt_bbox: List
+        :return: return percent of salient pixel overlap with the ground truth
+        :rtype: Float
+        """
+        good = 0
+        total = 0
+        total_percent = 0
+        for iindex, i in enumerate(saliency_scores):
+            for jindex, j in enumerate(i):
```
Comment (Advitya17) [inline python/vision_explanation_methods/explanation_evaluation/pointing_game.py:149]: For the for loop starting at 149, where `i` is a tensor, I'm curious if `good` & `total` can be calculated with torch operations.

Something like
```python
good += torch_operations(i)
total += torch_operations2(i)
```

Curious if it may be faster due to torch's optimizations (even though you're using a single for loop here and the torch operation approach may be going through each tensor twice).

Could explore if have bandwidth on how many such math operations / metrics computations can be offloaded to torch. Happy to assist as needed.

File: tests/test_vision_explanation.py [hunk 2]
```diff
@@ -193,3 +195,39 @@ def test_vision_explain_loadmodel():
...
+    _ = download_assets(modelpath)
+
+    # run the main function for saliency map generation
+    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+    model = _get_instance_segmentation_model(5)
+    model.load_state_dict(torch.load(modelpath, device))
+    model.to(device)
+    model.eval()
+    detection_model = PytorchDRiseWrapper(model=model,
+                                          number_of_classes=1000)
```
Comment (Advitya17) [inline tests/test_vision_explanation.py:220]: Curious, why is this set to 1000?

File: tests/test_vision_explanation.py [hunk 3]
```diff
@@ -193,3 +195,39 @@ def test_vision_explain_loadmodel():
...
+
+    # find saliency scores for top 20% of salient pixels
+    # do this for the second object in an image
+    pg = PointingGame(detection_model)
+    s = pg.pointing_game(imgpath, 1, .8)
```
Comment (Advitya17) [inline tests/test_vision_explanation.py:225]: Nit: maybe use more descriptive variable names for `s` & `v`

File: notebooks/pointing_game_example.ipynb [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import os\n",
+    "\n",
+    "import PIL.Image as Image\n",
+    "\n",
+    "import torchvision.models.detection as d\n",
+    "\n",
+    "import PIL.Image as Image\n",
```
Comment (imatiach-msft) [inline notebooks/pointing_game_example.ipynb:15]: PIL.Image is imported twice here (and line 13)

File: notebooks/pointing_game_example.ipynb [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    "from ml_wrappers.model.image_model_wrapper import (PytorchDRiseWrapper)\n",
```
Comment (imatiach-msft) [inline notebooks/pointing_game_example.ipynb:16]: I think there are no parens needed here around (PytorchDRiseWrapper)

File: notebooks/pointing_game_example.ipynb [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    "\n",
+    "from vision_explanation_methods.explanation_evaluation.pointing_game import PointingGame"
```
Comment (imatiach-msft) [inline notebooks/pointing_game_example.ipynb:18]: explanation_evaluation seems quite long and is sort of duplicated by the "explanation" part of "vision_explanation_methods" top-level namespace.  What do you think about just evaluation?

File: python/vision_explanation_methods/common/constants.py [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+"""Defines constants for vision-explanation-methods."""
+
+from enum import Enum
+
+
+class Device(Enum):
+    """Specifies all possible device types."""
+
+    CPU = 'cpu'
```
Comment (imatiach-msft) [inline python/vision_explanation_methods/common/constants.py:13]: I wonder if this can be re-imported from https://github.com/microsoft/ml-wrappers/blob/d83115785f50aef4e627dc1eba92d9c042f7c11e/python/ml_wrappers/common/constants.py#L319 ?
Hmm, or perhaps not since this repo does not have a dependency on ml-wrappers.  Actually, it seems it does have a dependency here:
https://github.com/microsoft/vision-explanation-methods/blob/main/python/setup.py#LL44C6-L44C17
so perhaps indeed it might be easier to just re-import here and reduce duplication.

File: python/vision_explanation_methods/explanation_evaluation/pointing_game.py [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+        If device were set to "auto", then the desired device will be cuda
+        (GPU) if available. Otherwise, the device should be set to cpu.
+
+        :param device: parameter specifying the device to move the model
+            to.
+        :type device: str
+        :return: selected device to run computations on
+        :rtype: str
+        """
+        if (device in [member.value for member in Device]
```
[the remark this answers is not in the document]
Comment (natalie-isak) [reply python/vision_explanation_methods/explanation_evaluation/pointing_game.py:172]: good point!!

File: python/vision_explanation_methods/explanation_evaluation/pointing_game.py [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+        total_percent = 0
+        for iindex, i in enumerate(saliency_scores):
+            for jindex, j in enumerate(i):
+                if j > 0:
+                    if (gt_bbox[0] < iindex < gt_bbox[2]
+                       and gt_bbox[1] < jindex < gt_bbox[3]):
+                        good += 1
+                    total += 1
+        percent = good/total
+        total_percent += percent
```
[the remark this answers is not in the document]
Comment (natalie-isak) [reply python/vision_explanation_methods/explanation_evaluation/pointing_game.py:156]: lol - not sure what I was thinking there. Must have been leftover code from an old version, thx!

File: tests/test_vision_explanation.py (hunk 3 above)
Comment (natalie-isak) [reply tests/test_vision_explanation.py:225]: don't see a v, but resolved for s

File: python/vision_explanation_methods/common/constants.py (hunk 7 above)
Comment (natalie-isak) [reply python/vision_explanation_methods/common/constants.py]: good point!

File: python/vision_explanation_methods/evaluation/pointing_game.py [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+        temp[temp < threshold] = -1
+        return temp
+
+    def visualize_highly_salient_pixels(self,
+                                        img,
+                                        saliency_scores):
+        """
+        Create figure of highly salient pixels.
+
+        :param imagelocation: Path of the image location
```
Comment (imatiach-msft) [inline python/vision_explanation_methods/evaluation/pointing_game.py:105]: nit: img not imglocation based on params above

File: tests/test_vision_explanation.py (hunk 3 above)
Comment (imatiach-msft) [inline tests/test_vision_explanation.py:225]: Can the 1 and .8 be moved to some variables?  Is there a recommended default threshold that could be used?  It seems like that might be .8?

File: python/vision_explanation_methods/evaluation/pointing_game.py [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+
+        positive_mask = saliency_scores > 0
+        positive_gt_mask = torch.logical_and(positive_mask, gt_mask)
+
+        good = positive_gt_mask.sum().item()
+        total = positive_mask.sum().item()
+
+        return good / total
+
+    def _get_device(self, device: str) -> str:
```
[the remark this answers is not in the document]
Comment (imatiach-msft) [reply python/vision_explanation_methods/evaluation/pointing_game.py:159]: just want to minimize code duplication across repos

[the remark this answers is not in the document]
Comment (imatiach-msft) [reply python/vision_explanation_methods/evaluation/pointing_game.py:159]: perhaps we should eventually make this a public method in ml-wrappers repo then

File: python/vision_explanation_methods/evaluation/pointing_game.py [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+        :type saliency_scores: List[Tensor]
+        :param gt_bbox: bounding box for ground truth prediction
+        :type gt_bbox: List
+        :return: return percent of salient pixel overlap with the ground truth
+        :rtype: Float
+        """
+        saliency_scores = torch.tensor(saliency_scores)
+        gt_bbox = torch.tensor(gt_bbox)
+
+        gt_mask = torch.zeros_like(saliency_scores, dtype=torch.bool)
```
Comment (Advitya17) [inline python/vision_explanation_methods/evaluation/pointing_game.py:154]: Looks great, thanks a lot for all the efforts here!

File: python/vision_explanation_methods/evaluation/pointing_game.py [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+        gt_mask[gt_bbox[0]:gt_bbox[2], gt_bbox[1]:gt_bbox[3]] = True
+
+        positive_mask = saliency_scores > 0
```
Comment (Advitya17) [inline python/vision_explanation_methods/evaluation/pointing_game.py:157]: Nit: Presume this can also be done in torch, dk if that may be faster under the hood

**Review by Advitya17 — APPROVED** [review]

File: python/vision_explanation_methods/evaluation/pointing_game.py (hunk 13 above)
Comment (natalie-isak) [reply python/vision_explanation_methods/evaluation/pointing_game.py:157]: The latency difference between using torch.gt() and creating a boolean mask using a comparison operator (>) will depend on various factors, including the size of the input tensors and the hardware on which the computations are performed.

In general, torch.gt() and the comparison operator will have similar performance characteristics because torch.gt() is implemented as a low-level operation using efficient tensor operations. However, the exact latency difference can vary based on the specific hardware, memory layout, and other factors.

It's worth noting that creating a boolean mask using a comparison operator (>) may involve additional memory allocation for the resulting boolean tensor, which can slightly impact the overall latency compared to using torch.gt() directly.

In practice, if you are working with large tensors or performing computationally intensive operations, the performance difference between torch.gt() and creating a boolean mask using a comparison operator is likely to be negligible. It is recommended to choose the approach that aligns better with the overall code structure and readability rather than solely focusing on potential latency differences.


Source: ChatGPT :)

File: python/vision_explanation_methods/explanation_evaluation/pointing_game.py [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+        :return: selected device to run computations on
+        :rtype: str
+        """
+        if (device in [member.value for member in Device]
+           or type(device) == int or device is None):
+            if device == Device.AUTO.value:
+                return (Device.CUDA.value if torch.cuda.is_available()
+                        else Device.CPU.value)
+            return device.value
+        else:
```
[the remark this answers is not in the document]
Comment (Advitya17) [reply python/vision_explanation_methods/explanation_evaluation/pointing_game.py]: It'll automatically get out of the condition and can raise the ValueError, `else` won't be needed as there's nothing else in this function definition.

File: python/vision_explanation_methods/evaluation/pointing_game.py (hunk 13 above)
Comment (Advitya17) [reply python/vision_explanation_methods/evaluation/pointing_game.py:157]: If I understand correctly, using torch.gt can be better in terms of memory of the entire, though may be by a small margin in practice. Another benefit could be that the entire metric could be calculated on a GPU rather than combining intermediate computations from GPU & CPU.

May encourage the below update just coz it's a trivial update into a principally better logic.

```suggestion
        positive_mask = torch.gt(saliency_scores, 0)
```

**Review by Advitya17 — APPROVED** [review]

File: python/vision_explanation_methods/explanation_evaluation/pointing_game.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (Advitya17) [reply python/vision_explanation_methods/explanation_evaluation/pointing_game.py]: It'll automatically get out of the condition and can raise the ValueError, else won't be needed as there's nothing else in this function definition.

File: python/vision_explanation_methods/evaluation/pointing_game.py (hunk 13 above)
Comment (Advitya17) [reply python/vision_explanation_methods/evaluation/pointing_game.py:157]: If I understand correctly, torch.gt may be slightly better in terms of memory of the tensor, though perhaps in practice the difference could still be minimal. Another benefit could be the entire metric would be calculated on GPU instead of combining intermediate computations from CPU & GPU.

Would encourage the below update as it's a trivial update into a principally better logic
```suggestion
        positive_mask = torch.gt(saliency_scores, 0)
```

File: python/vision_explanation_methods/evaluation/pointing_game.py [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+        positive_gt_mask = torch.logical_and(positive_mask, gt_mask)
+
```
Comment (Advitya17) [inline python/vision_explanation_methods/evaluation/pointing_game.py:159]: Nit
```suggestion
good = positive_gt_mask.sum()
total = positive_mask.sum()

return torch.div(good, total).item()
```

**Review by Advitya17 — APPROVED** [review]

File: python/vision_explanation_methods/evaluation/pointing_game.py (hunk 13 above)
Comment (natalie-isak) [reply python/vision_explanation_methods/evaluation/pointing_game.py:157]: Sure, my interpretation was that the difference was negligible as size increased and hardware changed, but your point about offering faster computation for GPU is valid. Will commit the change.

File: python/vision_explanation_methods/evaluation/pointing_game.py (hunk 15 above)
Comment (natalie-isak) [reply python/vision_explanation_methods/evaluation/pointing_game.py:159]: nice catch!

File: python/vision_explanation_methods/explanation_evaluation/pointing_game.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (natalie-isak) [reply python/vision_explanation_methods/explanation_evaluation/pointing_game.py]: comment is now outdated -- now importing this function from ml-wrappers. @Advitya17 I am closing this comment for now but will discuss offline with you whether or not we need to open an issue on ml-wrappers
